content/client-lib-development-guide/features.textile

```diff
@@ -287,9 +292,10 @@ h3(#realtime-connection). Connection
 * @(RTN6)@ A @Connection@ is successful and considered @CONNECTED@ once the websocket connection is open and the intial @CONNECTED@ @ProtocolMessage@ has been received
 * @(RTN19)@ If the @CONNECTED@ @ProtocolMessage@ contains a @connectionDetails@ property, the attributes within @ConnectionDetails@ will be used as the defaults for this client library, overriding any configured options at the time the @CONNECTED@ @ProtocolMessage@ is received
 * @(RTN7)@ @ACK@ and @NACK@:
-** @(RTN7a)@ All @ProtocolMessage@ @Presence@ and @Message@ objects sent to Ably expect either an @ACK@ or @NACK@ from Ably to confirm successful receipt and acceptance or failure respectively
+** @(RTN7a)@ All @ProtocolMessage@ @Presence@ and @Message@ objects sent to Ably expect either an @ACK@ or @NACK@ from Ably to confirm successful receipt and acceptance or failure respectively. For clarity, it is unnecessary to fail the publish operation of a message using a timer. Instead the client library can rely on the realtime system to send an @ACK@ or @NACK@ when connected, to fail all awaiting messages when @DISCONNECTED@ for too long as described in "RTN7d":#RTN7d, or upon reconnecting, the corresponding @ACK@ or @NACK@ protocol messages will be received as described in "RTN19a":#RTN19a
 ** @(RTN7b)@ Every @ProtocolMessage@ sent must contain a unique serially incrementing @msgSerial@ integer value starting at zero. The @msgSerial@ along with the @count@ for incoming @ACK@ and @NACK@ @ProtocolMessages@ indicates which messages succeeded or failed to be delivered
 ** @(RTN7c)@ If a connection enters the @CLOSED@ or @FAILED@ state, or if the connection state is lost, and a @ACK@ or @NACK@ has not yet been received for a message, the client should consider the delivery of those messages as failed
+** @(RTN7d)@ If a connection enters the @DISCONNECTED@ or @SUSPENDED@ state, and has not become @CONNECTED@ within 15 seconds, then the client should consider the delivery of those messages as failed
```

Comment:
> @mattheworiordan I'm sorry, I'm getting really confused here. AFAICT, this is an argument against nacking-after-15s for messages that have already been attempted once, and in favour of nacking-after-15s for messages which haven't been attempted yet. But the spec item (RTN7d) we're commenting on only applies to messages which have been attempted (RTN7) and are awaiting ack/nack.

Well no, I don't think so.  Lets put queued messages aside for now and focus only on messages published but awaiting an `ACK` / `NACK`.

If you publish a message, bearing in mind we're realtime system, then if the message does not get an `ACK` / `NACK` in a sensible amount of time, we should IMHO notify the developer so that they can decide what to do next. Whilst it is possible that the message was indeed published, holding on indefinitely or for a very long time feels wrong because a) the message may not have been published, b) the UI probably needs updating to state it's probably not sent, c) there is no guarantee we can confirm `ACK` / `NACK` correctly anyway once they've been disconnected for a longer period of time.

So I think our design should aim to:

* Hide any transient network failures from the developer such as changing networks, or a very brief disconnection.  This is going to cover most situations where a connection becomes disconnected
* Fail requests that are not processed in a reasonable amount of time (15s in mind is more than enough)
* When in doubt, prefer a failure notification as opposed to success / silence (we may need to add this to the docs for message publishing)

In regards to messages that have not yet been attempted and are sitting in the queue, in fact I would concur that we should probably apply the same 15s timeout rule and fail immediately if the connection becomes `suspended`, `closed` or `failed` (which we do anyway).   For example, if the connection is in the `DISCONNECTED` state and remains in that state for 2 minutes, all queued messages will just sit there waiting to be sent.  If we try and maintain the premise that if we cannot deliver soon then it's considered a failure, we should really time these messages out after 15s.

@paddybyers WDYT?